src/goto-harness/goto_harness_parse_options.cpp

```diff
@@ -7,46 +7,168 @@ Author: Diffblue Ltd.
+    << " goto-harness --version           show version\n"
+    << " goto-harness <in> <out> --harness-function-name <name> --harness-type "
+       "<harness-type> [harness options]\n"
+    << "\n"
+    << "<in> <out>                 goto binaries to read from / write to\n"
+    << "--harness-function-name    the name of the harness function to "
+       "generate\n"
+    << "--harness-type             one of the harness types listed below\n"
+    << "\n\n"
+    << FUNCTION_HARNESS_GENERATOR_HELP << eom;
 }
 
 goto_harness_parse_optionst::goto_harness_parse_optionst(
   int argc,
   const char *argv[])
   : parse_options_baset{GOTO_HARNESS_OPTIONS, argc, argv, ui_message_handler},
+    messaget(ui_message_handler),
     ui_message_handler(cmdline, std::string("GOTO-HARNESS ") + CBMC_VERSION)
 {
 }
+
+goto_harness_parse_optionst::goto_harness_configt
+goto_harness_parse_optionst::handle_common_options()
+{
+  goto_harness_configt config{};
+
+  // This just checks the positional arguments to be 2.
+  // Options are not in .args
+  if(cmdline.args.size() != 2)
+  {
+    help();
+    throw invalid_command_line_argument_exceptiont{
+      "need to specify both input and output goto binary file names (may be "
+      "the same)",
+      "<in goto binary> <out goto binary>"};
+  }
+
+  config.in_file = cmdline.args[0];
+  config.out_file = cmdline.args[1];
+
+  if(!cmdline.isset(GOTO_HARNESS_GENERATOR_TYPE_OPT))
+  {
+    throw invalid_command_line_argument_exceptiont{
+      "required option not set", "--" GOTO_HARNESS_GENERATOR_TYPE_OPT};
+  }
+  config.harness_type = cmdline.get_value(GOTO_HARNESS_GENERATOR_TYPE_OPT);
+
+  // Read the name of the harness function to generate
+  if(!cmdline.isset(GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT))
+  {
+    throw invalid_command_line_argument_exceptiont{
+      "required option not set",
+      "--" GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT};
+  }
+  config.harness_function_name = irep_idt{
+    cmdline.get_value(GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT)};
+
+  return config;
+}
+
+goto_harness_generator_factoryt goto_harness_parse_optionst::make_factory()
+{
+  auto factory = goto_harness_generator_factoryt{};
+  factory.register_generator("call-function", [this]() {
+    return util_make_unique<function_harness_generatort>(ui_message_handler);
+  });
+  return factory;
+}
+
+goto_harness_generator_factoryt::generator_optionst
+goto_harness_parse_optionst::collect_generate_factory_options()
+{
+  auto const common_options =
+    std::set<std::string>{"version",
+                          GOTO_HARNESS_GENERATOR_TYPE_OPT,
+                          GOTO_HARNESS_GENERATOR_HARNESS_FUNCTION_NAME_OPT};
+
+  auto factory_options = goto_harness_generator_factoryt::generator_optionst{};
+
+  for(auto const &option : cmdline.option_names())
+  {
+    if(common_options.find(option) == common_options.end())
+    {
+      factory_options.insert({option, cmdline.get_value(option.c_str())});
```

Comment:
Ah. I misunderstood how `get_value()` works with repeated arguments :(

Have to reconsider this part of the API as well :+1:

Reply:
We're doing something somewhat similar to what you suggested now; `handle_option` is now always taking a list of values. Instead of an overload like you suggested we've added helper functions `require_exactly_one_value`, which returns the only value of a single value list or throws if it is not given a single value list, and `assert_no_values` which asserts that it is given an empty list (assert because an option which doesn't have values _can't_ get values as per the cmdline API).